logging/src/lib.rs

```diff
@@ -234,57 +233,56 @@ impl<CB: ContainerBuilder, T> std::ops::Deref for TypedLogger<CB, T> {
     }
 }
 
-impl<CB: ContainerBuilder, A: ?Sized + FnMut(&Duration, &mut CB::Container)> LoggerInner<CB, A> {
+impl<CB: ContainerBuilder, A: ?Sized + FnMut(&Duration, &mut Option<CB::Container>)> LoggerInner<CB, A> {
     fn log_many<I>(&mut self, events: I)
         where I: IntoIterator, CB: PushInto<(Duration, I::Item)>,
     {
         let elapsed = self.time.elapsed() + self.offset;
         for event in events {
-            self.dirty = true;
             self.builder.push_into((elapsed, event.into()));
             while let Some(container) = self.builder.extract() {
-                (self.action)(&elapsed, container);
+                let mut c = Some(std::mem::take(container));
+                (self.action)(&elapsed, &mut c);
+                if let Some(mut c) = c {
+                    c.clear();
+                    *container = c;
+                }
             }
         }
     }
 
     fn flush(&mut self) {
         let elapsed = self.time.elapsed() + self.offset;
 
-        let mut action_ran = false;
-        while let Some(buffer) = self.builder.finish() {
-            (self.action)(&elapsed, buffer);
-            action_ran = true;
-        }
-
-        if !action_ran {
-            // Send an empty container to indicate progress.
-            (self.action)(&elapsed, &mut CB::Container::default());
+        while let Some(container) = self.builder.finish() {
+            let mut c = Some(std::mem::take(container));
+            (self.action)(&elapsed, &mut c);
+            if let Some(mut c) = c {
+                c.clear();
+                *container = c;
+            }
         }
 
-        self.dirty = false;
+        // Send no container to indicate flush.
+        (self.action)(&elapsed, &mut None);
```

Comment:
A nit, but one of the reasons to take a `&mut Option<_>` rather than a `Option<&mut _>` is to allow the `None` call to pass back resources, and at least with `Push` the intent is that you keep calling this as long as you get a non-`None` back. We don't have to do that here, and probably massively over-thinking this, but wanted to call out the gap.

Reply:
I see your point, but I'm wondering if this is actually true! Or put differently, is it something we'd _like_ to be true, or is it something that we had at some point and forgot about?

The reason I'm asking is because most (all?) places where we call `Push::done`, we don't have a loop to drain resources. If I recall correctly, the only place where we loop is in Differential merge batchers to drain the stash of allocations once we're done merging chains.

We could change `Push`'s `done` function to look like this instead:
```
    fn done(&mut self) { 
        let mut container = None;
        loop {
            self.push(&mut container);
            if container.is_none() { break; }
        }
    }
```